docs/doc/60-contributing/03-rfcs/20220512-versioned-config.md

```diff
@@ -0,0 +1,146 @@
+---
+title: Versioned Config
+description: RFC for versioned config
+---
+
+- RFC PR: [datafuselabs/databend#5324](https://github.com/datafuselabs/databend/pull/5324)
+- Tracking Issue: [datafuselabs/databend#5297](https://github.com/datafuselabs/databend/issues/5297)
+
+# Summary
+
+The versioned configuration will allow us to iterate quickly while avoiding breaking the environment.
+
+# Motivation
+
+While early birds are starting to deploy databend by themselves, it's time for us to establish some contracts between users. We should allow users to upgrade their deployments without breaking backward compatibility. In this RFC, we will focus on `config`.
+
+`config` I mentioned here including:
+
+- config file that is read by `databend-query` and `databend-meta`.
+- config env that read by `databend-query` and `databend-meta`
+- application args that are accepted by `databend-query` and `databend-meta`
+- protobuf messaged that generated by `databend-query` (stored inside `databend-meta`)
+
+Out of scope:
+
+- Tools like `fuzz` and `metactl` are not covered by this RFC.
+- Command-line UX of `databend-query` and `databend-meta` is another topic. We will not cover it in this RFC.
+- Config input/output by SQL/HTTP Rest API are not covered (for example, the output of table `system.config`)
+
+> For convenience, I will use `databend` to refer to `databend-query` and `databend-meta`.
+
+With this RFC, our users will upgrade their deployments without breaking. Old configs should always work along with the new implementations.
+
+# Guide-level explanation
+
+With this RFC, users will have a new field called `version` at the top of the config file. The value of the version is an integer and defaults to `0`. If the version is not specified, we will load it as` 0`.
+
+All config files, env, and args will be supported:
+
+- config file: `version=42`
+- config env: `export CONFIG_VERSION=42`
```

Comment:
what'd happen when the versions between the config env and config file got conflicts?

Reply:
There are no conflicts. As they are the same field, `env` will always overwrite `file`. (The same rules we apply for all fields.)

Reply:
Well, I came up with some ideas like config file and env are using a different version, can they by loaded at the same time.

It's possible, but a bit complex. I expect to support the overwrite behavior first.

Will be added in `Future possibilities` sections.

Reply:
> Well, I came up with some ideas like config file and env are using a different version, can they by loaded at the same time?

oh i'd hope this not happen, it's a bit anti-intuitive, and hard to reason about which field really acts in effect